File: Data Structures & Algorithms/concatenation-of-array/submission-2.cpp
class Solution {
public:
    vector<int> getConcatenation(vector<int>& nums) {
        vector<int> answer(2 * nums.size());
        
        for (int i = 0; i < nums.size()*2; i++){
            answer[i] = nums[i%(nums.size())];
        }

        return answer;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # for constant space just store the prev, and current of left and right respectively and use that to calculate the next ones
        left = []
        init_right = []
        length = len(nums)

        for index in range(length):
            if index == 0:
                left.append(nums[index])
                init_right.append(nums[length-(index+1)])
            else:
                left.append(nums[index] * left[index-1])
                init_right.append(nums[length-(index+1)] * init_right[index-1])

        right = init_right[::-1]

        res = []
        for index, num in enumerate(nums):
            if index == 0:
                res.append(right[1])
                continue
            if index == (len(nums) - 1):
                res.append(left[index-1])
                continue
            res.append(left[index-1] * right[index + 1])

        return res

